beacon-chain/db/kv/state.go

```diff
@@ -99,30 +114,112 @@ func (s *Store) SaveStates(ctx context.Context, states []iface.ReadOnlyBeaconSta
 		return errors.New("nil state")
 	}
 	multipleEncs := make([][]byte, len(states))
+	validatorsEntries := make(map[string][]byte)              // It's a map to make sure that you store only new validator entries.
+	validatorKeys := make([][]byte, len(states))              // For every state, this stores a compressed list of validator keys.
+	realValidators := make([][]*ethpb.Validator, len(states)) // It's temporary structure to restore state in memory after persisting it.
 	for i, st := range states {
 		pbState, err := v1.ProtobufBeaconState(st.InnerStateUnsafe())
 		if err != nil {
 			return err
 		}
+		if featureconfig.Get().EnableHistoricalSpaceRepresentation {
+			// store the real validators to restore the in memory state later.
+			realValidators[i] = pbState.Validators
+
+			// yank out the validators and store them in separate table to save space.
+			var hashes []byte
+			for _, val := range pbState.Validators {
+				valBytes, encodeErr := encode(ctx, val)
+				if encodeErr != nil {
+					return encodeErr
+				}
+
+				// create the unique hash for that validator entry.
+				hash := hashutil.Hash(valBytes)
+				hashes = append(hashes, hash[:]...)
+
+				// note down the hash and the encoded validator entry
+				hashStr := hexutils.BytesToHex(hash[:])
```

Comment:
why make it hex here ? you can simply cast the byte slice to a string if that is what is needed.

Reply:
`hexString` vs simple string differ in size. Anyway i went for simple string for simplicity